File: internal/quality/quality.go
package quality

import (
	"context"
	"path/filepath"
	"strings"
	"time"

	"moyuan-code/internal/fsutil"
	"moyuan-code/internal/logging"
	"moyuan-code/internal/process"
	"moyuan-code/internal/workspace"
)

type Check struct {
	Type     string  `json:"type"`
	Command  *string `json:"command"`
	Started  string  `json:"started_at,omitempty"`
	Finished string  `json:"finished_at,omitempty"`
	Status   string  `json:"status"`
	ExitCode *int    `json:"exit_code,omitempty"`
	Stdout   string  `json:"stdout,omitempty"`
	Stderr   string  `json:"stderr,omitempty"`
	Reason   string  `json:"reason,omitempty"`
}

type Report struct {
	ID        string  `json:"id"`
	TaskID    string  `json:"task_id"`
	CreatedAt string  `json:"created_at"`
	Status    string  `json:"status"`
	Checks    []Check `json:"checks"`
}

func Run(ctx context.Context, rootDir string, taskID string) (Report, error) {
	ws, err := workspace.Load(rootDir)
	if err != nil {
		return Report{}, err
	}
	reportID := "quality-" + taskID + "-" + time.Now().UTC().Format("20060102150405")
	checks := []Check{}
	checks = appendChecks(ctx, rootDir, checks, "build", ws.Project.Stack.BuildCommands)
	checks = appendChecks(ctx, rootDir, checks, "lint", ws.Project.Stack.LintCommands)
	checks = appendChecks(ctx, rootDir, checks, "test", ws.Project.Stack.TestCommands)
	status := "passed"
	for _, check := range checks {
		if check.Status == "failed" {
			status = "failed"
			break
		}
	}
	report := Report{
		ID:        reportID,
		TaskID:    taskID,
		CreatedAt: time.Now().UTC().Format(time.RFC3339Nano),
		Status:    status,
		Checks:    checks,
	}
	paths := reportPaths(rootDir, reportID)
	if err := fsutil.WriteJSON(paths.json, report); err != nil {
		return Report{}, err
	}
	md := []string{"# Quality Report", "", "- Task ID: `" + taskID + "`", "- Report ID: `" + reportID + "`", "- Status: `" + status + "`", ""}
	for _, check := range checks {
		line := "- " + check.Type + ": " + check.Status
		if check.Command != nil {
			line += " (" + *check.Command + ")"
		}
		md = append(md, line)
	}
	if err := fsutil.WriteText(paths.md, strings.Join(md, "\n")+"\n"); err != nil {
		return Report{}, err
	}
	_ = fsutil.AppendJSONL(filepath.Join(workspace.ForRoot(rootDir).QualityDir, "events.jsonl"), map[string]any{
		"ts":        time.Now().UTC().Format(time.RFC3339Nano),
		"task_id":   taskID,
		"report_id": reportID,
		"status":    status,
	})
	_ = logging.Log(rootDir, "quality", "quality.check.completed", map[string]any{"task_id": taskID, "report_id": reportID, "status": status})
	return report, nil
}

func Read(rootDir string, reportID string) (Report, bool, error) {
	var report Report
	found, err := fsutil.ReadJSON(reportPaths(rootDir, reportID).json, &report)
	return report, found, err
}

func appendChecks(ctx context.Context, rootDir string, checks []Check, typ string, commands []string) []Check {
	if len(commands) == 0 {
		return append(checks, Check{Type: typ, Status: "skipped", Reason: "no " + typ + " command configured"})
	}
	for _, command := range commands {
		started := time.Now().UTC().Format(time.RFC3339Nano)
		result := process.RunShell(ctx, rootDir, command)
		finished := time.Now().UTC().Format(time.RFC3339Nano)
		status := "passed"
		if result.Code != 0 {
			status = "failed"
		}
		code := result.Code
		cmd := command
		checks = append(checks, Check{
			Type:     typ,
			Command:  &cmd,
			Started:  started,
			Finished: finished,
			Status:   status,
			ExitCode: &code,
			Stdout:   result.Stdout,
			Stderr:   result.Stderr,
		})
	}
	return checks
}

func reportPaths(rootDir string, id string) struct{ json, md string } {
	base := filepath.Join(workspace.ForRoot(rootDir).QualityDir, "reports")
	return struct{ json, md string }{
		json: filepath.Join(base, id+".json"),
		md:   filepath.Join(base, id+".md"),
	}
}

File: src/core/quality.js
import path from 'node:path';
import { appendJsonl, fileExists, readJson, readYaml, writeJson, writeText } from '../lib/fs.js';
import { getWorkspacePaths } from './paths.js';
import { runShell } from '../lib/process.js';
import { logEvent } from './logging.js';

function reportPaths(rootDir, id) {
  const base = getWorkspacePaths(rootDir).qualityDir;
  return {
    json: path.join(base, 'reports', `${id}.json`),
    md: path.join(base, 'reports', `${id}.md`)
  };
}

async function getCommandLists(rootDir) {
  const project = await readYaml(getWorkspacePaths(rootDir).projectYaml, null);
  const stack = project?.stack ?? {};
  return {
    build: stack.build_commands ?? [],
    lint: stack.lint_commands ?? [],
    test: stack.test_commands ?? []
  };
}

async function runChecks(rootDir, commandList, type) {
  const results = [];
  for (const command of commandList) {
    const startedAt = new Date().toISOString();
    const result = await runShell(command, { cwd: rootDir });
    results.push({
      type,
      command,
      started_at: startedAt,
      finished_at: new Date().toISOString(),
      status: result.code === 0 ? 'passed' : 'failed',
      exit_code: result.code,
      stdout: result.stdout,
      stderr: result.stderr
    });
  }
  return results;
}

export async function runQualityCheck(rootDir, taskId, options = {}) {
  const commands = await getCommandLists(rootDir);
  const reportId = options.reportId ?? `quality-${taskId}-${Date.now()}`;
  const paths = reportPaths(rootDir, reportId);

  const checks = [];
  if (commands.build.length) {
    checks.push(...(await runChecks(rootDir, commands.build, 'build')));
  } else {
    checks.push({ type: 'build', command: null, status: 'skipped', reason: 'no build command configured' });
  }
  if (commands.lint.length) {
    checks.push(...(await runChecks(rootDir, commands.lint, 'lint')));
  } else {
    checks.push({ type: 'lint', command: null, status: 'skipped', reason: 'no lint command configured' });
  }
  if (commands.test.length) {
    checks.push(...(await runChecks(rootDir, commands.test, 'test')));
  } else {
    checks.push({ type: 'test', command: null, status: 'skipped', reason: 'no test command configured' });
  }

  const failed = checks.some((check) => check.status === 'failed');
  const report = {
    id: reportId,
    task_id: taskId,
    created_at: new Date().toISOString(),
    status: failed ? 'failed' : 'passed',
    checks
  };

  await writeJson(paths.json, report);
  await writeText(
    paths.md,
    [
      `# Quality Report`,
      '',
      `- Task ID: \`${taskId}\``,
      `- Report ID: \`${reportId}\``,
      `- Status: \`${report.status}\``,
      '',
      ...checks.map((check) => `- ${check.type}: ${check.status}${check.command ? ` (${check.command})` : ''}`)
    ].join('\n') + '\n'
  );

  await appendJsonl(path.join(getWorkspacePaths(rootDir).qualityDir, 'events.jsonl'), {
    ts: new Date().toISOString(),
    task_id: taskId,
    report_id: reportId,
    status: report.status
  });

  await logEvent(rootDir, 'quality', 'quality.check.completed', {
    task_id: taskId,
    report_id: reportId,
    status: report.status
  });

  return report;
}

export async function readQualityReport(rootDir, reportId) {
  const paths = reportPaths(rootDir, reportId);
  if (!(await fileExists(paths.json))) {
    return null;
  }
  return readJson(paths.json, null);
}
